hta/common/trace_filter.py

```diff
@@ -0,0 +1,363 @@
+from abc import ABC, abstractmethod
+from typing import List, Optional, Tuple, Union
+
+import pandas as pd
+
+from hta.common.trace import logger, TraceSymbolTable
+from hta.utils.utils import get_symbol_column_names
+
+
+class Filter(ABC):
+    """An abstract base class for trace event filters.
+
+    All concrete classes derived from Filter must implement the method __call__().
+    The general user patter is like follows:
+    1. Create a filter object.
+        filter_func = ConcreateFilter(...)
+    2. Apply the filter to an input DataFrame.
+        df = filter_func(df)
+    or an input DataFrame with the corresponding Trace Symbol Table.
+        df = filter_func(df, symbol_table)
+    """
+
+    @abstractmethod
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        """
+        Args:
+            df (pd.DataFrame): an input DataFrame.
+            symbol_table (TraceSymbolTable, optional): a trace symbol table for encoding/decoding the name
+                and cat columns of the input DataFrame.
+
+        Returns:
+            The output DataFrame.
+        """
+        raise NotImplementedError()
+
+
+class IterationFilter(Filter):
+    """A filter class for extracting the events of one or a list of iteration numbers.
+
+    An iteration number is equivalent to a profiler step number or iteration id. For instance, An IterationFilter
+     whose iterations attribute is [550,] will extract all trace events corresponding to the iteration 550.
+
+    Attributes:
+        iterations (int or List[int]): One or a list of iteration numbers for filtering the DataFrame.
+
+    Examples:
+        filter_func = IterationFilter(551)
+        selected_df = filter_func(df)
+        filter_func = IterationFilter([551, 552])
+        selected_df = filter_func(df)
+    """
+
+    def __init__(self, iterations: Union[int, List[int]]) -> None:
+        """
+        Initialize an IterationFilter object with specified iterations.
+
+        Args:
+            iterations (int or List[int]): One or a list of integers representing iteration numbers.
+        """
+        if not isinstance(iterations, (int, list)):
+            raise TypeError("Iterations must be an integer or a list of integers.")
+
+        self.iterations: List[int] = (
+            [iterations] if isinstance(iterations, int) else iterations
+        )
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        if "iteration" not in df.columns:
+            logger.warning("The DataFrame has no column named 'iteration'.")
+            return df
+
+        return df.loc[df["iteration"].isin(self.iterations)]
+
+
+class IterationIndexFilter(Filter):
```

Comment:
If multiple traces are available, which rank is used? Is there a default rank value?